### PR TITLE
Added unit test for pkg/cloud/services/autoscaling package

### PR DESCRIPTION
**Pull request by Madhur97**
**What type of PR is this?**
/area testing

**What this PR does / why we need it**:
This PR increases the unit tests coverage for `pkg/cloud/services/autoscaling` package.
Coverage increased from 13.7% to 94.2%

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #2910

**Special notes for your reviewer**:
Please add missing test scenarios [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2910) for tracking.

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Add unit tests for `pkg/cloud/services/autoscaling` package
```

### CODE REVIEW

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go [hunk 1]
```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
...
+		expect          func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder)
+		check           func(err error)
+	}{
+		{
+			name:            "should return without error if update ASG is successful",
+			machinePoolName: "update-asg-success",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				m.UpdateAutoScalingGroup(gomock.AssignableToTypeOf(&autoscaling.UpdateAutoScalingGroupInput{})).Return(&autoscaling.UpdateAutoScalingGroupOutput{}, nil)
+			},
+			check: func(err error) {
```
Comment (Ankitasw) [inline pkg/cloud/services/autoscaling/autoscalinggroup_test.go:573]: Can we refactor this check func as it is used in multiple places, or directly use it while checking [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/9a86ec7d19650f754cb20882f9b5baa3df802529/pkg/cloud/services/autoscaling/autoscalinggroup_test.go#L552)? Also, its usage is same across UpdateASG, CreateASG tests etc., so you can think of some refactoring

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go [hunk 2]
```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
...
+			},
+			check: func(autoscalinggroup *expinfrav1.AutoScalingGroup, err error) {
+				if err == nil {
+					t.Fatalf("expected error to occur")
+				}
+				if autoscalinggroup != nil {
+					t.Fatalf("expected asg to be nil but got: %+v", autoscalinggroup)
+				}
+			},
+		},
```
Comment (Ankitasw) [inline pkg/cloud/services/autoscaling/autoscalinggroup_test.go:536]: Can we also cover the following scenarios:
- should fail in getting subnets for ASG
- should fail if AWSMachinePool has no LaunchTemplateID

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go [hunk 3]
```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
...
+			machinePoolName: "update-asg-fail",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				m.UpdateAutoScalingGroup(gomock.AssignableToTypeOf(&autoscaling.UpdateAutoScalingGroupInput{})).Return(nil, awserrors.NewFailedDependency("dependency failure"))
+			},
+			check: func(err error) {
+				if err == nil {
+					t.Fatalf("expected error to occur")
+				}
+			},
+		},
```
Comment (Ankitasw) [inline pkg/cloud/services/autoscaling/autoscalinggroup_test.go:590]: Can we also add scenario:
Should fetch subnets from infra cluster if no subnets specified in AWSMachinePool

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go [hunk 4]
```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
...
+			mps := &scope.MachinePoolScope{
+				AWSMachinePool: &expinfrav1.AWSMachinePool{},
+			}
+			mps.AWSMachinePool.Name = tt.machinePoolName
+			out, err := s.CanStartASGInstanceRefresh(mps)
+			tt.check(out, err)
+		})
+	}
+}
+
```
Comment (Ankitasw) [inline pkg/cloud/services/autoscaling/autoscalinggroup_test.go:992]: Can we also add tests for `StartASGInstanceRefresh` func?
Also, can you check if `createSDKMixedInstancesPolicy` func can be covered in any of the flows?

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go [hunk 5]
```diff
@@ -18,20 +18,25 @@ package asg
```
Comment (Ankitasw) [inline pkg/cloud/services/autoscaling/autoscalinggroup_test.go:18]: If [BuildTags](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/9a86ec7d19650f754cb20882f9b5baa3df802529/pkg/cloud/services/autoscaling/autoscalinggroup.go#L394) func is not used anywhere, can we delete it?

**Review by Ankitasw — CHANGES_REQUESTED** [review]

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go (hunk 5 above)
Comment (Ankitasw) [reply pkg/cloud/services/autoscaling/autoscalinggroup_test.go:18]: @sedefsavas can you comment here if this is ok?

**Comment by Ankitasw** [comment]
@Madhur97  You can also check past issues occurred due to issues in autoscaling and cover those as well in these tests(like integration tests)

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go (hunk 5 above)
Comment (sedefsavas) [reply pkg/cloud/services/autoscaling/autoscalinggroup_test.go:18]: Yes, let's cleanup unused methods.

**Comment by sedefsavas** [comment]
@Madhur97  did you resolve the review comments with the latest push? Is this PR ready for another review?

**Comment by Ankitasw** [comment]
@Madhur97 did you rebase your changes? looks like a lot of change in recent commit.

File: pkg/cloud/services/autoscaling/autoscalinggroup_test.go [hunk 6]
```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
...
+			tt.check(out, err)
+		})
+	}
+}
+
+func getMachinePoolScope(machinePoolName string) (*scope.MachinePoolScope, error) {
+	scheme := runtime.NewScheme()
+	_ = infrav1.AddToScheme(scheme)
+	_ = expinfrav1.AddToScheme(scheme)
+	client := fake.NewClientBuilder().WithScheme(scheme).Build()
```
[the remark this answers is not in the document]
Comment (Madhur97) [reply pkg/cloud/services/autoscaling/autoscalinggroup_test.go]: We cannot pass nil due to this check [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/scope/machinepool.go#L77)

**Comment by Madhur97** [comment]
> @Madhur97 did you resolve the review comments with the latest push? Is this PR ready for another review?

@sedefsavas Now its ready for another review.

**Comment by Ankitasw** [comment]
Thanks @Madhur97 for resolving the comments.
/lgtm
/assign @sedefsavas 
for her comments review.

**Comment by Ankitasw** [comment]
Gentle reminder for this as well

> @Madhur97 You can also check past issues occurred due to issues in autoscaling and cover those as well in these tests(like integration tests)

**Comment by Madhur97** [comment]
> Gentle reminder for this as well
> 
> > @Madhur97 You can also check past issues occurred due to issues in autoscaling and cover those as well in these tests(like integration tests)

Added in the parent [issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2910#issuecomment-1028782971) for future reference.

**Pull request merged**